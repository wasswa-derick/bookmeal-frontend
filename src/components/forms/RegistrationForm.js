import React from "react";
import validator from "validator";
import { Link } from "react-router-dom";
import PropTypes from "prop-types";
import FormInput from "./FormInput";

/**
 * @class RegistrationForm
 * @extends {React.Component}
 */
class RegistrationForm extends React.Component {
  state = {
    data: {
      name: "",
      email: "",
      password: ""
    },
    errors: {}
  };

  /**
   * @param {Event} evt
   *@returns {null} sets value of target element to the corresponding state value
   */
  onChange = evt => {
    this.setState({
      data: { ...this.state.data, [evt.target.name]: evt.target.value }
    });
  };
  /**
   *@param {Event} evt
   * @returns {null} returns null
   */
  onFormSubmit = evt => {
    evt.preventDefault();
    const { data } = this.state;
    let errors = this.validateData(data);

    if (Object.keys(errors).length === 0) {
      this.props.handleSubmit(data).catch(err => {
        switch (err.response.status) {
          case 500:
            errors.form = "Internal server error occured";
            break;
          case 400:
            errors = { ...err.response.data.errors };
            break;
          default:
            break;
        }
        this.setState({ errors });
      });
    }

    this.setState({ errors });
  };
  /**
   *@param {Object} data
   * @returns {Object} errors object
   */
  validateData = data => {
    const errors = {};
    if (validator.isEmpty(data.email)) {
      errors.email = "This field is required";
    } else if (!validator.isEmail(data.email)) {
      errors.email = "This email is invalid";
    }
    if (validator.isEmpty(data.password)) {
      errors.password = "This field is required";
    }
    if (validator.isEmpty(data.name)) {
      errors.name = "This field is required";
    }
    return errors;
  };

  /**
   * @returns {form} for registration
   */
  render() {
    const { data, errors } = this.state;
    return (
      <form autoComplete="off" onSubmit={this.onFormSubmit}>
        <FormInput
          value={data.name}
          type="text"
          name="name"
          label="Full Name"
          onChange={this.onChange}
          error={errors.name}
        />
        <FormInput
          name="email"
          type="email"
          label="Email"
          value={data.email}
          onChange={this.onChange}
          error={errors.email}
        />
        <FormInput
          name="password"
          type="password"
          label="Password"
          value={data.password}
          onChange={this.onChange}
          error={errors.password}
        />
        <button type="submit" className="btn btn-primary btn-submit">
          Create Account
        </button>
        <p>
          Already have an account?
          <Link href="/login" to="/login">
            Log In
          </Link>
        </p>
      </form>
    );
  }
}

RegistrationForm.propTypes = {
  handleSubmit: PropTypes.func.isRequired
};

export default RegistrationForm;
