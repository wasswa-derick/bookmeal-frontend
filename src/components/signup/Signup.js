import React from "react";
import { connect } from "react-redux";
import PropTypes from "prop-types";
import { RegistrationForm } from "../common/forms";
import { registerCustomer } from "../../actions/auth";
/**
 * @class Signup
 * @extends {React.Component}
 */
export class Signup extends React.Component {
  /**
   * @param {object} data to submit
   * @returns {Promise} promise to the caller
   * @memberof Signup
   */
  handleSubmit = data =>
    this.props
      .registerCustomer(data)
      .then(() => this.props.history.push("/login"));

  /**
   * @returns {any} renders a page
   * @memberof Signup
   */
  render() {
    return (
      <div>
        <div className="container login-panel">
          <div className="card auth-card">
            <div className="card-body">
              <h3 className="card-title title">Create your Account</h3>
              <RegistrationForm handleSubmit={this.handleSubmit} />
            </div>
          </div>
        </div>
      </div>
    );
  }
}

Signup.propTypes = {
  registerCustomer: PropTypes.func.isRequired,
  history: PropTypes.shape({
    push: PropTypes.func.isRequired
  }).isRequired
};
export default connect(null, { registerCustomer })(Signup);
