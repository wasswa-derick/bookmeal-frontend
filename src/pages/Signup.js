import React from "react";
import RegistrationForm from "../components/forms/RegistrationForm";

const Signup = () => (
  <div className="container login-panel">
    <div className="card auth-card">
      <div className="card-body">
        <h3 className="card-title title">Create your Account</h3>
        <RegistrationForm />
      </div>
    </div>
  </div>
);

export default Signup;
