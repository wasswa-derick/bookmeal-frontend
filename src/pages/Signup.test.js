import React from "react";
import { shallow } from "enzyme";
import Signup from "./Signup";
import RegistrationForm from "../components/forms/RegistrationForm";

describe("Signup Page", () => {
  let wrapper;
  beforeEach(() => {
    wrapper = shallow(<Signup />);
  });

  it("should render registration form", () => {
    expect(wrapper.containsMatchingElement(<RegistrationForm />)).toEqual(true);
  });
});
