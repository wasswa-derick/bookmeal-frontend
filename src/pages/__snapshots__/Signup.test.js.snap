// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Signup Page should render correctly 1`] = `
<div>
  <Connect(Component) />
  <div
    className="container login-panel"
  >
    <div
      className="card auth-card"
    >
      <div
        className="card-body"
      >
        <h3
          className="card-title title"
        >
          Create your Account
        </h3>
        <RegistrationForm
          handleSubmit={[Function]}
        />
      </div>
    </div>
  </div>
</div>
`;
